services/bonita-time-tracker/src/main/java/org/bonitasoft/engine/tracking/FlushThread.java

```diff
@@ -20,40 +20,57 @@
 
     private final TimeTracker timeTracker;
 
-    private final long flushIntervalInMilliSeconds;
-
-    private final Clock clock;
-
     private final TechnicalLoggerService logger;
 
-    public FlushThread(final Clock clock, final long flushIntervalInMilliSeconds, final TimeTracker timeTracker, final TechnicalLoggerService logger) {
-        super("TimeTracker-FlushThread");
-        this.clock = clock;
-        this.logger = logger;
-        this.flushIntervalInMilliSeconds = flushIntervalInMilliSeconds;
+    public FlushThread(final TimeTracker timeTracker) {
+        super("Bonita-TimeTracker-FlushThread");
+        this.logger = timeTracker.getLogger();
         this.timeTracker = timeTracker;
     }
 
     @Override
     public void run() {
         info("Starting " + this.getName() + "...");
+        long lastFlushTimestamp = System.currentTimeMillis();
         while (true) {
+            final long now = System.currentTimeMillis();
             try {
-                clock.sleep(flushIntervalInMilliSeconds);
+                final long sleepTime = getSleepTime(now, lastFlushTimestamp);
+                if (this.logger.isLoggable(getClass(), TechnicalLogSeverity.DEBUG)) {
+                    this.logger.log(getClass(), TechnicalLogSeverity.DEBUG, "FlushThread: sleeping for: " + sleepTime + "ms");
+                }
+                this.timeTracker.getClock().sleep(sleepTime);
             } catch (InterruptedException e) {
                 break;
             }
-            try {
-                this.timeTracker.flush();
-            } catch (Exception e) {
-                if (this.logger.isLoggable(getClass(), TechnicalLogSeverity.WARNING)) {
-                    this.logger.log(getClass(), TechnicalLogSeverity.WARNING, "Exception caught while flushing: " + e.getMessage(), e);
-                }
-            }
+            lastFlushTimestamp = flush(now);
         }
         info(this.getName() + " stopped.");
     }
 
+    long getSleepTime(final long now, final long lastFlushTimestamp) throws InterruptedException {
+        final long flushDuration = now - lastFlushTimestamp;
+        return this.timeTracker.getFlushIntervalInMS() - flushDuration;
+    }
+
+    long flush(final long now) {
+        try {
+            final FlushResult flushResult = this.timeTracker.flush();
+            return flushResult.getFlushTime();
```

Comment:
it's strange to delegate to the flush result what is the last flush timestamp, should the flush result return the time to flush and let the thread determine when was it's last flush timestamp